keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
+5. The first cidr entry of `--services-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default address family.
+6. Attempting to allocate `ClusterIP` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--services-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
 
 #### Type NodePort, LoadBalancer, ClusterIP
-Services will only be assigned an address from a single address family (either IPv4 or IPv6). The address family for the Service’s cluster IP is determined by setting the field, `.spec.ipFamily`, on that Service. You can only set this field when creating a new Service. Setting the `.spec.ipFamily` field is optional and should only be used if you plan to enable IPv4 and IPv6 Services and Ingresses on your cluster. The configuration of this field not a requirement for egress traffic. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller-manager. The `.spec.ipFamily` field can be set to either IPv4 or IPv6.
+In a dualstack cluster Services can be:
+1. Singlestack: assigned from the first or the second cidr entries as set using `--services-cluster-ip-range` flag, in this scenario Services will have a single allocated `ClusterIP`
+2. dualstack: assigned from both the first and the second cidr entries, in this scenario Services will have a dual assigned `ClusterIPs`
+
+The above is achieved using the following changes to Service api.
+1. Add `spec.ipFamilyPolicy` field with possible values of  `SingleStack`, `DualStack`, or `nil`.
+2. Add a `spec.ipFamilies[]` field with possible values of `IPv4`, `IPv6` for each each entry or `nil` for the entire array. Maximum entries in this array is two and must be `IPv4`, `IPv6` in any order.
+3. Add a `spec.ClusterIPs[]` field that follows the same rules as `status.PodIPs/status.PodIP`. Maximum entries in this array is two entries of two different ip address families. Semantically `spec.ClusterIP` will be referred as Service Primary IP.
+
+The rules that govern these fields are described in the below sections
+
+##### Creating a New Single Stack Service
+
+User who want to create a new single stack Service can provide `nil` values for both `spec.ipFamilyPolicy` and `spec.ipFamilies`. The results of this are:
+1. the new Service will be defaulted to `spec.ipFamilyPolicy: SingleStack`.
```

Comment:
I understand. But shouldn't we at least set it before storing the service, or at allocation time? otherwise any user will have to look at `.spec.ClusterIPs` and count them in order to understand if that service is dual stack or not which sounds to me like a a suboptimal experience.

Reply:
You have to count `clusterIPs` / `ipFamilies` anyway; `ipFamilyPolicy` is a request, not a requirement. (Specifically, in the case of `ipFamilyPolicy: DualStack` in a single-stack cluster.)

Reply:
What @danwinship said, I think.  The policy is really only for the allocator - real consumers of Services should look at `ipFamilies`

Reply:
> > * "I don't care at all (nil, nil, nil)"
> 
> Well... it _might_ mean "I don't care", but it might also mean "I care but I am a legacy Service and I don't know how to tell you what I care about".
 
That is a distinction without a difference, I think.  :)

> Really, it's "(DualStack, nil, nil)" that means "I don't care" (in the sense of, "I will function correctly regardless of what `ipFamilies` you assign me, so just give me the families that make the most sense for this cluster")

I think it's not importnt, but I disagree.  (nil, nil, nil) means I have no
preference at all.  (DualStack, nil, nil) means I prefer dual-stack if it is
available.

I think that distinction is academic, though.

> I'm wondering if the "mutating webhook to set `ipFamilyPolicy: DualStack`" idea really makes sense. It basically requires you to manually modify all of your legacy obligatorily-single-stack Services to specify `ipFamilyPolicy: SingleStack` to avoid them being mutated... but if you're going to have to manually modify a bunch of services, why not just manually modify all of your happy-to-be-dual-stack Services to specify `ipFamilyPolicy: DualStack` instead? (Particularly in the case of "third party" Service definitions that the admin/user didn't create themselves, the upstream project or whatever will presumably want to modify the service to have `ipFamilyPolicy: DualStack` anyway, to work correctly in dual-stack clusters that _don't_ have a mutating admission hook.)

You can tell the difference between a Create and and Update in the admission
control.  There should be no need to bulk-change anything.

> If `ipFamilyPolicy` is defined the way it is in order to support a theoretical admission hook, then we need to explicitly describe how we're expecting that to work, so we can make sure it actually makes sense...
 
Yeah, we should probably just write one as an example.

> > * "I care whether it is single or dual (value, nil, nil)"
> 
> Actually, when does `ipFamily: SingleStack` make sense? The obvious reason for wanting a service to be obligatorily single-stack is that the pod is only listening on a single IP family. But in that case, you can't say `ipFamilyPolicy: SingleStack`, because it doesn't guarante you'll get the _right_ family. You have to say specifically either `ipFamilies: [ "IPv4" ]` or `ipFamilies: [ "IPv6" ]`.

Yes, it may not make a lot of sense, but the semantics of it are valid and
consistent.

> So... (policy, families, ips):
> - I'm not telling you what I want so you'd better make the safest possible assumption (single-stack of the default family) (nil, nil, nil)
> - I am able to do dual-stack, but I don't require it, and I don't care about IPs (DualStack, nil, nil)
> - ~I want single-stack but don't care which family (SingleStack, nil, nil)~
> - I can only do IPv4, but don't care about which IP (nil, [IPv4], nil).
> - I can only do IPv6, but don't care about which IP (nil, [IPv6], nil).
> - I am dual-stack but I have a specific preference about which family EndpointController and other single-stack controllers should use (nil, values, nil) (?)
> - I want specific IPs (nil, nil, value)"
 
Yes.  It seems to not have werid corner cases, which I rather like.

Reply:
> (nil, nil, nil) means I have no
> preference at all. (DualStack, nil, nil) means I prefer dual-stack if it is
> available.
> 
> I think that distinction is academic, though.

I agree that the distinction between "I have no preference at all" and "I prefer dual-stack if it is available" is academic.

I _disagree_ that (nil, nil, nil) means "I have no preference". There are many existing Service yaml files out there implicitly specifying (nil, nil, nil), which only support single-stack. If they had been created post-IPv6DualStack, they would have known to specify that they were single-stack. But they weren't, so they didn't. It's weird that newly-created services can indicate that they are single-stack, but we allow a cluster to blindly assume that legacy services aren't.

I guess the counter-argument is "well don't put a mutate-to-DualStack admission hook in the cluster unless you know all the legacy services are dual-stack ready", and that makes me wonder what sort of weird use case this is where the administrator somehow knows every Service that will ever be deployed, but for some reason can't just _manually_ edit those Services to specify `ipFamilyPolicy: DualStack` directly instead of relying on an admission hook to do it.

> > It basically requires you to manually modify all of your legacy obligatorily-single-stack Services to specify ipFamilyPolicy: SingleStack to avoid them being mutated...
>
> You can tell the difference between a Create and and Update in the admission
> control. There should be no need to bulk-change anything.

I meant for the case of creating a new service based on an old yaml definition. (Perhaps because you are deploying a third-party tool.) If you have a mutate-to-DualStack webhook, then you have to fix those legacy definitions to include `ipFamilyPolicy: SingleStack` before creating them so that they won't get mutated.

> Yes, it may not make a lot of sense, but the semantics of it are valid and
> consistent.

That's how NetworkPolicy ended up so confusing.

If `ipFamilyPolicy: SingleStack` doesn't make sense then we should think about whether it's possible to rearrange things in a way that gets rid of it.

Reply:
I'm operating under the assumption that the vast, vast majority of Services really will not care whether they are dual-stack or not, and of those services that only listen on one family, the vast vast majority aren't going to BREAK if the second family is present.  Am I wrong?

And in fact, we're talking about Services, not Pods.  If the app would break in the face of the other family, Service is far too late to be expressing that.

I think this is what @anguslees was saying in the doc - having services publish both families is exceedingly unlikely to ACTUALLY be a problem for anyone at all.  Our insistence on defaulting to single-family behavior is really out of paranoia and fear.

Pragmatically, it would be ugly to silently allocate a second clusterIP for existing services, so we won't do that.  We could silently convert headless services to dual-stack and just rely on happy-eyeballs to steer clients right, and I suspect we'd be fine.

After talking this through, I'm OK to embrace the paranoia, but let's call it what it is and ACK that it is making work that we *probably* don't strictly *need* to do.

Given that, it seems reasonable to empower people to bypass the paranoia and set DualStack by default.  IMO, that is more aligned with what *I* would expect to happen.

I am happy to re-arrange the schema to make silly-but-valid configurations go away, if you have a suggestion?  But we do need to make actual progress on this soon...

Reply:
> and of those services that only listen on one family, the vast vast majority aren't going to BREAK if the second family is present

To clarify, you are arguing that essentially all clients will correctly handle IPv6-to-IPv4 fallback, and therefore we don't have to worry about advertising dual-stack cluster IPs for IPv4-only services? Maybe that's true.

> Am I wrong?

I don't know...

> And in fact, we're talking about Services, not Pods. If the app would break in the face of the other family, Service is far too late to be expressing that.

Well, yes, we still have to solve the problem of how to do health checks of pods running single-stack servers in a dual-stack cluster (but maybe the answer there is just to fix the health check code to try both PodIPs).

> I am happy to re-arrange the schema to make silly-but-valid configurations go away, if you have a suggestion? But we do need to make actual progress on this soon...

Possibility 1: replace `ipFamilyPolicy: SingleStack` with `ipFamilyPolicy: IPv4` (or `SingleStackIPv4`?) and `ipFamilyPolicy: IPv6`. Except that then there's no difference between `ipFamilyPolicy: IPv4, ipFamilies: nil` and `ipFamilyPolicy: nil, ipFamilies: ["IPv4"]`. So that leads to...

Possibility 2: turn `ipFamilyPolicy` into `preferDualStack bool`, where `false` is exactly equivalent to `ipFamilyPolicy: nil` and `true` is exactly equivalent to `ipFamilyPolicy: DualStack`. Then you have:
- "I don't care at all (false, nil, nil)"
- "I prefer dual-stack (true, nil, nil)"
- "I care which families, but not which addresses (false, value, nil)"
  - ie, either "I am IPv4-only" or "I am IPv6-only" or "I want to be specifically dual-stack-IPv4-primary or dual-stack-IPv6-primary"
- "I care which IPs (false, nil, value)"

A dual-stack-ifying admission hook can change `preferDualStack: false` to `preferDualStack: true` on any Service where `ipFamilies` and `clusterIPs` are unset.

(Maybe we require `preferDualStack` to be `true` if either `ipFamilies` or `clusterIPs` has two value.)

Reply:
> To clarify, you are arguing that essentially all clients will correctly handle IPv6-to-IPv4 fallback, and therefore we don't have to worry about advertising dual-stack cluster IPs for IPv4-only services? Maybe that's true.

Yep, that's what I'm claiming.  Or at least, I feel we need sufficient actual examples to make us think this is an issue worth compromising our forever-API around.  So far I believe we have zero such examples.

Reply:
Two divergent points here.

@danwinship: replacing `ipFamilyPolicy` with `preferDualStack` seems OK to me.  It does eliminate the weird corner case, so kudos for rising to the challenge.

@anguslees I hear what you are saying here, and I am thinking really hard about it.  For existing Services that only have the historical single family, we're not going to quietly allocate them a second IP.  It feels weird to me that headless services would behave differently.  This is really a sticking point for me.

I think the semantics proposed here, perhaps modified by Dan's suggestion here are pretty simple and, when I spoke to customers about this, are "good enough".  It's a little more complicated and a little more paranoid, but the (admittedly small number of) customers and users I spoke to seemed to understand it.

@khenidak what do you think about Dan's proposed change?

Reply:
I have applied that change to kep